### DISCUSSION TITLE
Issue using `FillPath` with a recursion structure and hidden fields

### DISCUSSION THREAD

**Discussion by helderco** [cue-lang/cue#1676] [Q&A] [answered]
## Overview

In Dagger we have an action that reads a json/yaml document and converts leaf values into secrets (a struct with a reference to a value in memory). So consider the following example, decrypted from SOPS:

```yaml
FOO: bar
ONE:
    TWO:
        THREE: one-hundred-twenty-three
```

We want to convert the strings `bar` and `one-hundred-twenty-three` to secrets, while keeping their path to be referenced in other actions.

We've been using the current CUE type for that action's output:

```cue
output: #Secret | {[string]: output}
```

It has been working ok, but now we're seeing some issues when trying to use `FillPath`.

## Current scenario

Consider this repro, running in `cuelang.org/go v0.4.3`:

```go
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/format"
)

const source = `
#Secret: {
	$secret: _id: string
}

a: {
	// original scenario
	output: #Secret | {[string]: output}

	// attempted solution
	// output: _#out
	// _#out: #Secret | {[string]: _#out}

	// workaround to unblock
	// output: _
}

b: a.output.FOO
c: a.output.ONE.TWO.THREE
`

const output = `
FOO: $secret: _id: "100"
ONE: TWO: THREE: $secret: _id: "123"
`

const pkg = "core"

var ctx *cue.Context

func main() {
	ctx = cuecontext.New()

	src := NewValue(source)
	out := NewValue(output)

	// Debug("SOURCE", &src)

	final := src.FillPath(cue.ParsePath("a.output"), out)
	Debug("FINAL", &final)

	if final.Err() != nil {
		return
	}

	b := Lookup(final, "b")
	c := Lookup(final, "c")

	fmt.Printf("===> bId=%q; cId=%q\n", b, c)
}

func NewValue(s string) cue.Value {
	v := ctx.CompileString(s, cue.ImportPath(pkg))
	if v.Err() != nil {
		panic(v.Err())
	}
	return v
}

func Lookup(val cue.Value, field string) string {
	v := val.LookupPath(cue.MakePath(cue.Str(field), cue.Str("$secret"), cue.Hid("_id", pkg)))
	if !v.Exists() || !v.IsConcrete() {
		fmt.Printf("ERROR: %q: %v\n\n", field, v)
		return ""
	}
	s, err := v.String()
	if err != nil {
		panic(err)
	}
	return s
}

func Debug(label string, v *cue.Value, opts ...cue.Option) {
	b, err := format.Node(
		v.Eval().Syntax(opts...),
		format.UseSpaces(4),
		format.TabIndent(false),
	)
	if err == nil {
		fmt.Printf("%s\n%.*s\n%s\n\n", label, len(label), "================================", b)
	}
}
```

The output:

```
FINAL
=====
_|_ // a.output: 4 errors in empty disjunction: (and 4 more errors)
```

In our more complex setup we're seeing structural cycle issues, as seen in https://github.com/dagger/dagger/issues/1867:

```
actions.good.output.Password.Password: structural cycle
```
So it seems like keys are being repeated.

## Attempted solution

I [attempted a solution](https://github.com/dagger/dagger/pull/2024) ([changeset](https://github.com/dagger/dagger/compare/bb1e8e511dfa3490f7df9504ae315c6ffca759d8..c4dcc2db96e61e550448512c2d853587b540a747)) that doesn't reference back to `output`:

```cue
output: _#out
_#out: #Secret | {[string]: _#out}
```

But now we get empty disjunction issues:

```
FINAL
=====
{
    #Secret: {
        $secret: {}
    }
    a: {
        output: {
            FOO: {
                $secret: {}
            } | {
                $secret: {
                    $secret: {}
                } | {}
            }
            ONE: {
                TWO: {
                    THREE: {
                        $secret: {}
                    } | {
                        $secret: {
                            $secret: {}
                        } | {}
                    }
                }
            }
        }
    }
    b: {
        $secret: {}
    } | {
        $secret: {
            $secret: {}
        } | {}
    }
    c: {
        $secret: {}
    } | {
        $secret: {
            $secret: {}
        } | {}
    }
}

ERROR: "b": _|_ // field not found: $secret

ERROR: "c": _|_ // field not found: $secret

===> bId=""; cId=""
```

The issue, I think, is because the `_id` field, since it's hidden, it's not being enough to resolve the disjunction.

<details><summary><h4>It works if "id" is visible</h4></summary>

```diff
diff --git a/main.go b/main.go
index 72da1f4..df4cb43 100644
--- a/main.go
+++ b/main.go
@@ -10,7 +10,7 @@ import (
 
 const source = `
 #Secret: {
-	$secret: _id: string
+	$secret: id: string
 }
 
 a: {
@@ -30,8 +30,8 @@ c: a.output.ONE.TWO.THREE
 `
 
 const output = `
-FOO: $secret: _id: "100"
-ONE: TWO: THREE: $secret: _id: "123"
+FOO: $secret: id: "100"
+ONE: TWO: THREE: $secret: id: "123"
 `
 
 const pkg = "core"
@@ -68,7 +68,7 @@ func NewValue(s string) cue.Value {
 }
 
 func Lookup(val cue.Value, field string) string {
-	v := val.LookupPath(cue.MakePath(cue.Str(field), cue.Str("$secret"), cue.Hid("_id", pkg)))
+	v := val.LookupPath(cue.MakePath(cue.Str(field), cue.Str("$secret"), cue.Str("id")))
 	if !v.Exists() || !v.IsConcrete() {
 		fmt.Printf("ERROR: %q: %v\n\n", field, v)
 		return ""
```
Output:

```
FINAL
=====
{
    #Secret: {
        $secret: {
            id: string
        }
    }
    a: {
        output: {
            FOO: {
                $secret: {
                    id: "100"
                }
            }
            ONE: {
                TWO: {
                    THREE: {
                        $secret: {
                            id: "123"
                        }
                    }
                }
            }
        }
    }
    b: {
        $secret: {
            id: "100"
        }
    }
    c: {
        $secret: {
            id: "123"
        }
    }
}

===> bId="100"; cId="123"
```

</details>

## Workaround

To work around this, I can abandon the CUE definition and just use whatever's filled in runtime:

```cue
output: _
```

Which outputs correctly:

```
FINAL
=====
{
    #Secret: {
        $secret: {}
    }
    a: {
        output: {
            FOO: {
                $secret: {}
            }
            ONE: {
                TWO: {
                    THREE: {
                        $secret: {}
                    }
                }
            }
        }
    }
    b: {
        $secret: {}
    }
    c: {
        $secret: {}
    }
}

===> bId="100"; cId="123"
```

But that's not ideal.

## Question

So is there a way to define `output` that makes this structure work?

/cc @jlongtine

**Comment by myitcv** [comment]
Hi @helderco - I converted this to a discussion because I'm not clear whether there is a bug here or not, indeed it looks more like a question. 

If you do think there is a bug here, please can you clarify per the bug issue template what the expected vs observed behaviour is?

**Reply by helderco** [reply]
I also don't know if there's a bug or not. Just asking feedback wether that's the case or if there's a solution to this use case.

**Reply by myitcv** [reply]
Yep, understood. My suggestion would be to start with a discussion in these cases. That helps us keep issues "clean" - where there are clear bugs or feature requests.

**Comment by myitcv** [comment]
Analysis thus far. 

I've tried to extract what I believe to be the crux of the issue from your Go code:

```
exec cue eval x.cue

-- x.cue --
src: {
	#Secret: {
		$secret: _id: string
	}

	a: {
		output: #Secret | {[string]: output}
	}

}
out: {
	FOO: $secret: _id: "100"
	ONE: TWO: THREE: $secret: _id: "123"
}

final: src & {
	a: output: out
}
```

This gives:

```
final.a.output: 4 errors in empty disjunction:
final.a.output: field not allowed: FOO:
    ./x.cue:2:11
    ./x.cue:7:11
    ./x.cue:12:2
    ./x.cue:16:8
    ./x.cue:17:13
final.a.output.FOO: 2 errors in empty disjunction:
final.a.output.FOO: field not allowed: FOO:
    ./x.cue:2:11
    ./x.cue:7:11
    ./x.cue:7:32
    ./x.cue:12:2
    ./x.cue:12:7
    ./x.cue:16:8
    ./x.cue:17:13
final.a.output.FOO.FOO: structural cycle
```

Notice this removes fields `b` and `c` which don't appear to participate in the problem.

Will look to reduce this further.

**Reply by helderco** [reply]
Pretty good repro 🙂

**Reply by helderco** [reply]
I see the same results with the other attempts:

```
	a: {
		output: _#out
		_#out: #Secret | {[string]: _#out}
	}
```
and
```
	a: {
		output: _
	}
```

So it's not about `FillPath`, it's just CUE. Not sure I understand why we see this only when using this: https://github.com/dagger/dagger/blob/32cf54461cd6581b7dcb35156612debfc21efa6b/plan/plan.go#L63-L75 (which is what https://github.com/dagger/dagger/issues/1867 is reporting).

**Reply by myitcv** [reply]
> So it's not about `FillPath`, it's just CUE

Correct. I'm working on reducing things.

**Reply by myitcv** [reply]
Ok I think there are two issues going on here which makes it a bit harder. Here is a first cut at demonstrating one of those:

```
# case 1
exec cue eval -e o x.cue o1.cue
cmp stdout stdout.golden

# case 2
exec cue eval -e o x.cue o2.cue
cmp stdout stdout.golden

# case 3
exec cue eval -e o x.cue o3.cue
cmp stdout stdout.golden

-- x.cue --
package x

#Secret: {
	$secret: id: string
}

secrets: #Secret | {[string]: secrets}

src1: output: secrets
src2: output: #Secret | {[string]: output}

out: {
	FOO: $secret: id: "100"
	ONE: TWO: THREE: $secret: id: "123"
}
-- o1.cue --
package x

o: src1 & {
	output: out
}
-- o2.cue --
package x

o: output: src2.output & out
-- o3.cue --
package x

o: src2 & {
	output: out
}
-- stdout.golden --
output: {
    FOO: {
        $secret: {
            id: "100"
        }
    }
    ONE: {
        TWO: {
            THREE: {
                $secret: {
                    id: "123"
                }
            }
        }
    }
}
```

This should pass but instead gives:

```
# case 1 (0.018s)
# case 2 (0.016s)
# case 3 (0.015s)
> exec cue eval -e o x.cue o3.cue
[stderr]
o.output: 4 errors in empty disjunction:
o.output: field not allowed: FOO:
    ./o3.cue:3:4
    ./o3.cue:4:10
    ./x.cue:3:10
    ./x.cue:10:15
    ./x.cue:13:2
o.output.FOO: 2 errors in empty disjunction:
o.output.FOO: field not allowed: FOO:
    ./o3.cue:3:4
    ./o3.cue:4:10
    ./x.cue:3:10
    ./x.cue:10:15
    ./x.cue:10:36
    ./x.cue:13:2
    ./x.cue:13:7
o.output.FOO.FOO: structural cycle
[exit status 1]
FAIL: /tmp/testscript235354936/repro.txt/script.txt:10: unexpected command failure
```

**Reply by myitcv** [reply]
I'm going to try and work out the crux of the second issue and then see where we stand. But right now it looks like we'll have two issues to create off the back of this.

**Reply by helderco** [reply]
> So it's not about `FillPath`, it's just CUE

Going back, I tried to see the isse in CUE first. This was my first repro: https://cuelang.org/play/?id=-n61uazTm6r#cue@export@json

I saw the structural cycle error there and attempted a fix that solved it. The problem is that I used visible `id` fields in the playground that's why to me it appeared that the issue wasn't with CUE, but with FillPath so I went down that path.

With hidden `_id` fields, that disjunction isn't resolved that's why I saw a different result in practice.

**Reply by myitcv** [reply]
Indeed. I think there are two issues at play here, and you ran into both of them. Which made reasoning about things hard. I'll raise issues for both now. Thanks for helping to debug.

`FillPath` can always be rewritten in terms of pure CUE because it's just a convenience wrapper around a unification operation. 

I always try to reduce examples down to pure CUE if possible. The API is "just" a wrapper around evaluation of pure CUE evaluation.

**Reply by myitcv** [reply]
The two issues for reference: 

* https://github.com/cue-lang/cue/issues/1679
* https://github.com/cue-lang/cue/issues/1680

**Comment by helderco** [comment]
I think I found a way that works for us:

```
exec cue eval x.cue

-- x.cue --
src: {
	#Secret: {
		$secret: _id: string
	}

	a: {
		output: _#out
		_#out: #Secret | {[!~"\\$secret"]: _#out}
	}

}

out: {
	FOO: $secret: _id: "100"
	ONE: TWO: THREE: $secret: _id: "123"
}

final: src & {
	a: output: out
}
```
I just need to do a bit more testing.

**Reply by helderco** [reply]
I'm not being able to make this work in our environment. I'll need more time to get to the bottom of why, since it appears to work in CUE. I'm seeing the same thing as if `[!~"\\$secret"]` was just `[string]`.

**Reply by myitcv** [reply]
I'm not yet clear that this approach will work. We have an answer re https://github.com/cue-lang/cue/issues/1680, but we're not quite there on https://github.com/cue-lang/cue/issues/1679. 

Indeed regarding https://github.com/cue-lang/cue/issues/1679, is there any reason you use `_id` instead of `id`?

**Reply by helderco** [reply]
TLDR; They're hidden to protect against both visibility and manipulation by users.

Those `_id` are filled in Go (cue/flow) and they point to a value in a map (in memory). So when a new secret is created from a string (e.g., reading the contents of a file), an entry is filled in the map (`map[string]*Secret`), in this case (there's other types) with an id that is an hash of the secret's contents. Then when the secret needs to be used, we get that `_id` from CUE and pick the value from the map.

@jlongtine may explain better.

**Reply by jlongtine** [reply]
What @helderco said is correct. We're doing the same trick with our Dagger tasks (which we chatted about with you and @mpvl a while back) to hide the value in the `dagger.io` package. 

I see the note about lack of closedness with the hidden fields, which makes sense now that I think about it. We'll have to make sure that we're not using those in ways that could cause issues in the future.

**Reply by myitcv** [reply]
(I've just seen the full version of @helderco's response above - it looks like it was edited a few times after I read the first version that I received by email)

@jlongtine happy to explore potential solutions if the constraints have changed.

**Reply by myitcv** [reply]
@jlongtine per our conversation on Friday, we'd be happy to help explore potential solutions here but we'll need to do so in the context of the Dagger examples rather than a cut down repro.

**Comment by helderco** [comment, marked as answer]
Let me start fresh here with dagger's context.

I have a PR you can use to test these out, you can clone that and run the added test case (I'll provide commands):
- https://github.com/dagger/dagger/pull/2371

## 1. Structural cycle

As I said in the description, we had the following at first:

```cue
output: dagger.#Secret | {[string]: output}
```

It was reported that this **only happened** with [this FillPath](https://github.com/dagger/dagger/blob/8ea748fa12022233c51626ac62143a9d8cdb97e0/plan/plan.go#L81) in:
- https://github.com/dagger/dagger/issues/1867

## 2. Unresolved disjunction value

I attempted a fix with:

```cue
output: _#out
_#out: dagger.#Secret | {[string]: _#out}
```

It **solved the structural cycle** when using the `FillPath` above, but broke everywhere else where normal references were being used. 

That's because now ([with this test case](https://github.com/dagger/dagger/blob/3ed4d1f7658f8a3cf6098bcecb49ef90d7a9ba25/tests/tasks/decodesecret/decodesecret.cue)), we see that the `output` field has the following value:

```console
[actions.test.debug] ===> {
    _#def
    _#def: {
        FOO: {
            contents: {
                $dagger: {
                    secret: {}
                }
            } | {
                $dagger: {
                    secret: {
                        $dagger: {
                            secret: {}
                        }
                    } | {}
                }
            }
        }
        ONE: {
            TWO: {
                contents: {
                    $dagger: {
                        secret: {}
                    }
                } | {
                    $dagger: {
                        secret: {
                            $dagger: {
                                secret: {}
                            }
                        } | {}
                    }
                }
            }
        }
    }
}
...
[actions.test.verify.mounts.secret1.contents: source] ===> decode.output.FOO.contents
[actions.test.verify.mounts.secret1.contents: error] ===> actions.test.verify.mounts.secret1.contents: unresolved disjunction 
...
```

## 3. Unresolved disjunction error

To fix the disjunction, I attempted this then:

```cue
output: _#out
_#out: dagger.#Secret | {[!~"\\$dagger"]: _#out}
```

When debugging the `output` field, it **looks ok**, but not the [reference to a value](https://github.com/dagger/dagger/blob/3ed4d1f7658f8a3cf6098bcecb49ef90d7a9ba25/tests/tasks/decodesecret/decodesecret.cue#L54):

```console
➜  go run ./cmd/dagger/main.go do --log-format plain -p ./tests/tasks/decodesecret/decodesecret.cue test
[actions.test.debug] ===> {
    _#def
    _#def: {
        FOO: {
            contents: {
                $dagger: {
                    secret: {}
                }
            }
        }
        ONE: {
            TWO: {
                contents: {
                    $dagger: {
                        secret: {}
                    }
                }
            }
        }
    }
}
...
[actions.test.verify.mounts.secret1.contents: source] ===> decode.output.FOO.contents
[actions.test.verify.mounts.secret1.contents: error] ===> actions.test.verify.mounts.secret1.contents: unresolved disjunction
...
```

## 4. Working with FillPath?

BUT, if we use `FillPath` (from [CLI input](https://github.com/dagger/dagger/blob/8ea748fa12022233c51626ac62143a9d8cdb97e0/cmd/dagger/cmd/do.go#L149) here), then it works.

Notice `--format yaml`:

```console
➜  go run ./cmd/dagger/main.go do --log-format plain -p ./tests/tasks/decodesecret/decodesecret.cue test --format yaml
[actions.test.debug] ===> {
    _#def
    _#def: {
        FOO: {
            contents: {
                $dagger: {
                    secret: {}
                }
            }
        }
        ONE: {
            TWO: {
                contents: {
                    $dagger: {
                        secret: {}
                    }
                }
            }
        }
    }
}
...
[actions.test.verify.mounts.secret1.contents: source] ===> {
    _#def
    _#def: {
        $dagger: {
            secret: {}
        }
    }
}
[actions.test.verify.mounts.secret2.contents: source] ===> {
    _#def
    _#def: {
        $dagger: {
            secret: {}
        }
    }
}
...
```
Instead of `format` you can replicate with any field, not used anywhere. It's just the triggering of the code that does the `FillPath` that makes it work.

I don't know why, and not sure how to debug further. I'm using `_, args := v.Expr()` and getting the error with `args[0].Err()`.

<details><summary><h4>Full output (dagger do test)</h4></summary>

```console
➜  go run ./cmd/dagger/main.go do --log-format plain -p ./tests/tasks/decodesecret/decodesecret.cue test              
4:58PM INF actions.test.image | computing
4:58PM INF actions.test.write | computing
4:58PM INF actions.test.write | completed    duration=0s
4:58PM INF actions.test.secret | computing
4:58PM INF actions.test.secret | completed    duration=0s
4:58PM INF actions.test.decode | computing
4:58PM INF actions.test.decode | completed    duration=0s
4:58PM INF actions.test.debug | computing
[actions.test.debug] ===> {
    _#def
    _#def: {
        ONE: {
            TWO: {
                contents: {
                    $dagger: {
                        secret: {}
                    }
                }
            }
        }
        FOO: {
            contents: {
                $dagger: {
                    secret: {}
                }
            }
        }
    }
}
4:58PM INF actions.test.debug | completed    duration=0s
4:58PM INF actions.test.image | completed    duration=1.4s
4:58PM INF actions.test.verify | computing
[actions.test.verify.mounts.secret1.contents: source] ===> decode.output.FOO.contents
[actions.test.verify.mounts.secret1.contents: error] ===> actions.test.verify.mounts.secret1.contents: unresolved disjunction {ONE:{TWO:{contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 7477656c766500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 6 6 %!x(bool=false)}"}}}}},FOO:{contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 62617200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 3 3 %!x(bool=false)}"}}}},$dagger:{secret:{_id(dagger.io/dagger):string}}} | {ONE:{TWO:{contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 7477656c766500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 6 6 %!x(bool=false)}"}}}},$dagger:{secret:{_id(dagger.io/dagger):string}}} | {TWO:{contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 7477656c766500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 6 6 %!x(bool=false)}"}}},$dagger:{secret:{_id(dagger.io/dagger):string}}} | {contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 7477656c766500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 6 6 %!x(bool=false)}"}}} | {$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 7477656c766500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 6 6 %!x(bool=false)}"}}}}},FOO:{contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 62617200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 3 3 %!x(bool=false)}"}}},$dagger:{secret:{_id(dagger.io/dagger):string}}} | {contents:{$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 62617200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 3 3 %!x(bool=false)}"}}} | {$dagger:{secret:{_id(dagger.io/dagger):"&{[6a09e667 bb67ae85 3c6ef372 a54ff53a 510e527f 9b05688c 1f83d9ab 5be0cd19] 62617200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 3 3 %!x(bool=false)}"}}}}} (type struct) (and 1 more errors)
4:58PM ERR actions.test.verify | failed: invalid secret at path "actions.test.verify.mounts.secret1.contents": secret is not set
    duration=0s
4:58PM FTL system | failed to execute plan: task failed: actions.test.verify: invalid secret at path "actions.test.verify.mounts.secret1.contents": secret is not set


exit status 1
```

</details>
<details><summary><h4>Full output (dagger do test --format yaml)</h4></summary>

```console
➜  go run ./cmd/dagger/main.go do --log-format plain -p ./tests/tasks/decodesecret/decodesecret.cue test --format yaml
Changed: format: yaml
5:12PM INF actions.test.image | computing
5:12PM INF actions.test.write | computing
5:12PM INF actions.test.write | completed    duration=0s
5:12PM INF actions.test.secret | computing
5:12PM INF actions.test.secret | completed    duration=0s
5:12PM INF actions.test.decode | computing
5:12PM INF actions.test.decode | completed    duration=0s
5:12PM INF actions.test.debug | computing
[actions.test.debug] ===> {
    _#def
    _#def: {
        FOO: {
            contents: {
                $dagger: {
                    secret: {}
                }
            }
        }
        ONE: {
            TWO: {
                contents: {
                    $dagger: {
                        secret: {}
                    }
                }
            }
        }
    }
}
5:12PM INF actions.test.debug | completed    duration=0s
5:12PM INF actions.test.image | completed    duration=1.4s
5:12PM INF actions.test.verify | computing
[actions.test.verify.mounts.secret1.contents: source] ===> {
    _#def
    _#def: {
        $dagger: {
            secret: {}
        }
    }
}
[actions.test.verify.mounts.secret2.contents: source] ===> {
    _#def
    _#def: {
        $dagger: {
            secret: {}
        }
    }
}
5:12PM INF actions.test.verify | completed    duration=0s
```

</details>

## 5. Update: working config

This is working:

```cue
output: _#out
_#out: dagger.#Secret | *{[!~"\\$dagger"]: _#out}
```

Setting a default resolves the disjunction.

I just had a failing test in a non-task struct that I changed to:

```diff
-			token: testSecrets.output.NETLIFY_TOKEN.contents
+			token: testSecrets.output.NETLIFY_TOKEN.contents & dagger.#Secret
```

Thanks!

**Reply by helderco** [reply]
### Summary

1. https://github.com/dagger/dagger/pull/2371
   - Has [latest attempt to fix](https://github.com/dagger/dagger/blob/3ed4d1f7658f8a3cf6098bcecb49ef90d7a9ba25/pkg/dagger.io/dagger/core/secrets.cue#L23)
   - Has [test case](https://github.com/dagger/dagger/blob/3ed4d1f7658f8a3cf6098bcecb49ef90d7a9ba25/tests/tasks/decodesecret/decodesecret.cue)
   - Has commands to run test
      - `go run ./cmd/dagger/main.go do -p ./tests/tasks/decodesecret/decodesecret.cue test`
      - `go run ./cmd/dagger/main.go do -p ./tests/tasks/decodesecret/decodesecret.cue test --format yaml`
   - Has debugging code (look at the changes)
2. `FillPath` behavior
   - Structural cycle only with `FillPath`
   - Now latest fix only works with `FillPath`
      - `output` value looks ok, but referencing a field from it gives an _unresolved disjunction_ error
3. `FillPath` code:
   - When using `--with` flag:
      - https://github.com/dagger/dagger/blob/8ea748fa12022233c51626ac62143a9d8cdb97e0/plan/plan.go#L81
   - When using CLI input flags:
      - https://github.com/dagger/dagger/blob/8ea748fa12022233c51626ac62143a9d8cdb97e0/cmd/dagger/cmd/do.go#L149

Maybe @jlongtine can help debug further. I'm not sure what to try next. I'm going to try pairing with him.

**Reply by mpvl** [reply]
In reply to your sections:
1. Correct, that is indeed a structural cycle.
2. This is also correct. Both structs will match, as `[string]` matches `$secret` too. You can solve that by picking one of the structs as a default.
3. I'm not sure where the `$dagger` is coming from, so do don't know what else has changed. But using a regexp like that is indeed an alternative way to solve the issue from 2. I don't know how you printed these references. Too much seems to have changed from the original example vs this output. 
What are the exact steps to reproduce this? Especially, how do you construct this reference?

**Reply by helderco** [reply]
> I'm not sure where the `$dagger` is coming from, so do don't know what else has changed. But using a regexp like that is indeed an alternative way to solve the issue from 2. I don't know how you printed these references. Too much seems to have changed from the original example vs this output.
>        What are the exact steps to reproduce this? Especially, how do you construct this reference?

Everything you need to replicate is in the PR (https://github.com/dagger/dagger/pull/2371), I've posted links to the appropriate places. If you clone the repo and checkout the PR, you can run the provided commands and you should see the same output.

I'll copy stuff here for reference.

The actual definition for `#Secret` is this: 
```cue
#Secret: {
	$dagger: secret: _id: string
}
```

The test-case is the following:

```cue
package main

import (
	"encoding/json"
	"encoding/yaml"
	"dagger.io/dagger"
	"dagger.io/dagger/core"
)

let data = {
	FOO: "bar"
	ONE: TWO: "twelve"
}

dagger.#Plan & {
	actions: test: {
		// you can change this value by adding the `--format` flag to the `dagger do` command
		format: "json" | *"yaml"

		_formats: {
			"json": json.Marshal(data)
			"yaml": yaml.Marshal(data)
		}

		write: core.#WriteFile & {
			input:    dagger.#Scratch
			path:     "/secret"
			contents: _formats[format]
		}

		secret: core.#NewSecret & {
			input: write.output
			path:  "/secret"
		}

		decode: core.#DecodeSecret & {
			input:    secret.output
			"format": format
			// the `output` of #DecodeSecret is what we're having issues with
		}

		debug: core.#Nop & {
			// This is just for fmt.Printf the value of `decode.output`
			input: decode.output
		}

		image: core.#Pull & {
			source: "alpine"
		}

		verify: core.#Exec & {
			input: image.output

			mounts: {
				secret1: {
					type:     "secret"
					// `contents: dagger.#Secret` when `type: "secret"`
					contents: decode.output.FOO.contents
					dest:     "/secret1"
				}

				secret2: {
					type:     "secret"
					// second reference
					contents: decode.output.ONE.TWO.contents
					dest:     "/secret2"
				}
			}

			args: ["sh", "-c", "test $(cat /secret1) = bar && test $(cat /secret2) = twelve"]
		}
	}
}
```

### Debugging the `output` value
So the `#Nop` is where I debug the value of `#DecodeSecret.output`:

```go
	if src, err := v.Lookup("input").Source(); err == nil {
		fmt.Printf("[%s] ===> %s\n", v.Path(), src)
	}
```
Where [Source() does this](https://github.com/dagger/dagger/blob/c9780a90e0a2e30103176133e61e6a01a6bcbe16/compiler/value.go#L287-L296):
```go
import (
	...
	cueformat "cuelang.org/go/cue/format"
)

func (v *Value) Source(opts ...cue.Option) ([]byte, error) {
	...
	return cueformat.Node(v.val.Eval().Syntax(opts...),
		cueformat.UseSpaces(4),
		cueformat.TabIndent(false),
	)
}
```

### Debugging the references

As for the references, they [become](https://github.com/dagger/dagger/blob/c9780a90e0a2e30103176133e61e6a01a6bcbe16/pkg/dagger.io/dagger/core/exec.cue#L109-L115) `contents: decode.output.FOO.contents & dagger.#Secret`. I'm debugging them on the `switch` for the exec mounts:

```go
	case "secret":
		// -- start DEBUG --
		p := v.Lookup("contents").Path()
		_, args := v.Lookup("contents").Expr()

		if src, err := args[0].Source(); err == nil {
			fmt.Printf("[%s: source] ===> %s\n", p, src)
		}
		if err := args[0].Cue().Err(); err != nil {
			fmt.Printf("[%s: error] ===> %s\n", p, err)
		}
		// -- end DEBUG --

		// This is where it fails, in case the secret doesn't have a concrete `_id` value.
		contents, err := pctx.Secrets.FromValue(v.Lookup("contents"))
```

**Reply by helderco** [reply]
>  Both structs will match, as `[string]` matches `$secret` too. You can solve that by picking one of the structs as a default.

@mpvl Oh wow, this seems to be working!

```cue
_#out: dagger.#Secret | *{[!~"\\$dagger"]: _#out}
```
Running the test suite to confirm.

**Reply by helderco** [reply]
All tests are passing, thanks!

**Comment by mpvl** [comment]
Hmm, I'm not familiar with the `Cue` method. This doesn't seem to be the cue API.

**Reply by helderco** [reply]
We wrap cue values in a struct. The `Cue()` method just accesses the wrapped cue value:

```go

// Value is a wrapper around cue.Value.
// Use instead of cue.Value and cue.Instance
type Value struct {
	val cue.Value
	cc  *Compiler
}

func (v *Value) Cue() cue.Value {
	return v.val
}
```

https://github.com/dagger/dagger/blob/8ea748fa12022233c51626ac62143a9d8cdb97e0/compiler/value.go#L298-L300

**Reply by mpvl** [reply]
I would have expected that example not to get to a disjunction. It may be a bug.
